chore: update web vitals monitoring doc (#1796)
SigNoz/signoz.io: data/docs/frontend-monitoring/web-vitals-with-metrics.mdx

---
date: 2024-08-02
title: Web Vitals Monitoring with Metrics
id: web-vitals-with-metrics
---

This documentation provides a step-by-step guide to setting up web vitals monitoring with metrics using SigNoz and OpenTelemetry.

## Prerequisites

- [SigNoz Cloud](https://signoz.io/teams/) or self-hosted account
- A web application from where you want to send web vitals

## Setup

### Step 1: Setup OTel Collector
Install Otel collector binary using [these instructions](https://signoz.io/docs/tutorial/opentelemetry-binary-usage-in-virtual-machine/).

You would also need to update the collector config to whitelist the frontend domain. This is required to allow Cross-Origin Resource Sharing (CORS) requests from your frontend application to the OpenTelemetry collector.

```sh
receivers:
  otlp:
    protocols:
      grpc:
        endpoint: 0.0.0.0:4317
      http:
        endpoint: 0.0.0.0:4318
        cors:
          allowed_origins:
            - <<YOUR_FRONTEND_URL>>
          allowed_headers: ['*']
```

### Step 2: Install dependencies

<Tabs groupId="web-vitals-dependencies">

>   <TabItem value="npm" label="Npm">

Reviewer: ```suggestion
  <TabItem value="npm" label="Npm" default>
```

      npm install @opentelemetry/sdk-metrics               
      npm install @opentelemetry/resources                 
    ```
  </TabItem>
  <TabItem value="yarn" label="Yarn" default>
    ```sh
      yarn add web-vitals
      yarn add @opentelemetry/api
      yarn add @opentelemetry/exporter-metrics-otlp-http
      yarn add @opentelemetry/sdk-metrics               
      yarn add @opentelemetry/resources  
    ```
  </TabItem>
</Tabs>

<details>
<summary> Read more about the dependencies </summary>
- **`web-vitals`**: Core library for measuring and reporting web vitals metrics (LCP, INP, CLS, TTFB, FCP)
- **`@opentelemetry/api`**: Provides the core OpenTelemetry API for creating metrics and meters
- **`@opentelemetry/exporter-metrics-otlp-http`**: Exports metrics to OpenTelemetry Protocol (OTLP) over HTTP
- **`@opentelemetry/sdk-metrics`**: Provides the OpenTelemetry metrics SDK implementation
- **`@opentelemetry/resources`**: Defines service metadata and attributes for your application
</details>

### Step 3: Create an instrumentation file

The `instrument.ts` file is required to setup the `MeterProvider` which is used to create custom metrics.

> <Tabs groupId="web-vitals-instrumentation">

Reviewer: ```suggestion
<Tabs entityName="plans">
```

>   <TabItem value="cloud" label="SigNoz Cloud" default>

Reviewer: ```suggestion
  <TabItem value="signoz-cloud" label="SigNoz Cloud" default>
```

      'service.name': '<<SERVICE_NAME>>',
    });

    // Create a metric reader with OTLP exporter configured to send metrics to a local collector.
    const metricReader = new PeriodicExportingMetricReader({
      exporter: new OTLPMetricExporter({
        url: `https://ingest.<<INGESTION_REGION>>.signoz.cloud:443/v1/metrics`,
        headers: {
          'signoz-ingestion-key': <<INGESTION_KEY>>,
        },
      }),
      exportIntervalMillis: 10000, // Export metrics every 10 seconds.
    });

    // Initialize a MeterProvider with the above configurations.
    const myServiceMeterProvider = new MeterProvider({
      resource,
      readers: [metricReader],
    });

    // Set the initialized MeterProvider as global to enable metric collection across the app.
    metrics.setGlobalMeterProvider(myServiceMeterProvider);
    ```

    - Set the `<<INGESTION_REGION>>` to match your SigNoz Cloud [region](https://signoz.io/docs/ingestion/signoz-cloud/overview/#endpoint)
    - Replace `<<INGESTION_KEY>>` with your SigNoz [ingestion key](https://signoz.io/docs/ingestion/signoz-cloud/keys/)
    - `<<SERVICE_NAME>>` is the name of your service
  </TabItem>

>   <TabItem value="self-hosted" label="SigNoz Self-Hosted">

Reviewer: ```suggestion
  <TabItem value='self-host' label='Self-Host'>
```

</figure>

### Setup Dashboard and alerts
You can create a [custom dashboard](https://signoz.io/docs/userguide/manage-dashboards/#steps-to-create-a-custom-dashboard) and [set alerts](https://signoz.io/docs/setup-alerts-notification/)
to monitor your core web-vitals and be notified about anything critical.

You can find custom dashboard templates for web vitals [here](https://github.com/SigNoz/dashboards/tree/main/frontend-monitoring)

<figure data-zoomable align='center'>
<img className="box-shadowed-image" src="/img/events/launch-week-1/web-vitals.webp" alt="Core Web Vitals Monitoring in SigNoz"/>
<figcaption><i>Core Web Vitals Monitoring in SigNoz </i></figcaption>
</figure>

One of the cool feature of SigNoz charts is that you can create thresholds in the chart with different color coding. For example, in the below graph there are three thresholds for `good`, `needs improvement`, and `critical` level.

<figure data-zoomable align='center'>
<img className="box-shadowed-image" src="/img/events/launch-week-1/thresholds-chart.webp" alt="Thresholds in SigNoz"/>
<figcaption><i>Thresholds provide a visual cue about your application's performance </i></figcaption>
</figure>

> You can find and try the demo application [here](https://github.com/SigNoz/frontend-monitoring-examples/tree/main/web-vitals-with-metrics) and also use our [custom dashboards](https://github.com/SigNoz/frontend-monitoring-examples/tree/main/dashboard-templates).

Reviewer: ```suggestion

### Out of Box Dashboard
You can [import](https://signoz.io/docs/dashboards/import-dashboard/) our out-of-the-box [Dashboard for web vitals monitoring](https://github.com/SigNoz/frontend-monitoring-examples/tree/main/dashboard-templates)

## Demo Application

We have set up a [Demo React application](https://github.com/SigNoz/frontend-monitoring-examples/tree/main/web-vitals-with-metrics) that demonstrates monitoring Core Web Vitals along with metrics using OpenTelemetry and sending them to SigNoz.
```

---
date: 2024-08-02
title: Web Vitals Monitoring with Metrics
id: web-vitals-with-metrics
---

This documentation provides a step-by-step guide to setting up web vitals monitoring with metrics using SigNoz and OpenTelemetry.

## Prerequisites

- [SigNoz Cloud](https://signoz.io/teams/) or self-hosted account
- A web application from where you want to send web vitals

## Setup

> ### Step 1: Setup OTel Collector

Reviewer 2: This step is irrelevant as we are sending data directly to SigNoz endpoint

Reviewer 2: Same mistake has been made in other frontend monitoring docs, eg: - https://signoz-web-git-sig-2668-signoz.vercel.app/docs/frontend-monitoring/sending-logs/#step-1-setup-otel-collector

Author: We still need it for the self hosted version right?

      npm install @opentelemetry/sdk-metrics               
      npm install @opentelemetry/resources                 
    ```
  </TabItem>
  <TabItem value="yarn" label="Yarn">
    ```sh
      yarn add web-vitals
      yarn add @opentelemetry/api
      yarn add @opentelemetry/exporter-metrics-otlp-http
      yarn add @opentelemetry/sdk-metrics               
      yarn add @opentelemetry/resources  
    ```
  </TabItem>
</Tabs>

<details>
<summary> Read more about the dependencies </summary>
- **`web-vitals`**: Core library for measuring and reporting web vitals metrics (LCP, INP, CLS, TTFB, FCP)
- **`@opentelemetry/api`**: Provides the core OpenTelemetry API for creating metrics and meters
- **`@opentelemetry/exporter-metrics-otlp-http`**: Exports metrics to OpenTelemetry Protocol (OTLP) over HTTP
- **`@opentelemetry/sdk-metrics`**: Provides the OpenTelemetry metrics SDK implementation
- **`@opentelemetry/resources`**: Defines service metadata and attributes for your application
</details>

### Step 3: Create an instrumentation file

The `instrument.ts` file is required to setup the `MeterProvider` which is used to create custom metrics.

<Tabs entityName="plans">

>   <TabItem value="signoz-cloud" label="SigNoz Cloud" default>

Reviewer 2: No need to add tabs for SigNoz cloud and self hosted as difference is only 1-2 lines.
Just add a comment on right of header that for SigNoz self hosted header is not needed and also add a comment for endpoint

>     ```ts

Reviewer 2: Why only TS and no JS file?

Author: In this example, there were no differences between JS and TS code. So kept a single pane

---
date: 2024-08-02
title: Web Vitals Monitoring with Metrics
id: web-vitals-with-metrics
description: Set up up web vitals monitoring with metrics using SigNoz and OpenTelemetry

> tags: [Tech Tutorial, JavaScript, Frontend]

Reviewer 2: Remove tech tutorial, js, frontend tag, they don't exist

Reviewer 2: Just add SigNoz Cloud, Self-Host

---
date: 2025-11-05
title: Web Vitals Monitoring with Metrics
id: web-vitals-with-metrics

> description: Set up web vitals monitoring with metrics using SigNoz and OpenTelemetry

Reviewer 2: Update URL to opentelemetry-web-vitals-metrics
Update Title to Web Vitals Monitoring with OpenTelemetry Metrics
Update only titles, sidebar should stay same

tags: [SigNoz Cloud, Self-Host]
---

## Overview

> This documentation provides a step-by-step guide to setting up web vitals monitoring with metrics using SigNoz and OpenTelemetry.

Reviewer 2: This documentation provides a step-by-step guide to setting up web vitals monitoring with OpenTelemetry metrics on SigNoz.

</details>

### Step 2: Create an instrumentation file

The instrumentation file is required to setup the `MeterProvider` which is used to create custom metrics.

<Tabs entityName="plans">
  <TabItem value="ts" label="TypeScript" default>
    ```ts:instrumentation.ts
    import {
      MeterProvider,
      PeriodicExportingMetricReader,
    } from '@opentelemetry/sdk-metrics';
    import { OTLPMetricExporter } from '@opentelemetry/exporter-metrics-otlp-http';
    import { resourceFromAttributes } from '@opentelemetry/resources';
    import { metrics } from '@opentelemetry/api';

    // Define your resource, e.g., service name, environment.
    const resource = resourceFromAttributes({
      'service.name': '<SERVICE_NAME>',
    });

    // Create a metric reader with OTLP exporter configured to send metrics to a local collector.
    const metricReader = new PeriodicExportingMetricReader({
      exporter: new OTLPMetricExporter({
        // For the self-hosted version, please use the collector url instead.
        url: `https://ingest.<INGESTION_REGION>.signoz.cloud:443/v1/metrics`,
        headers: {

>           'signoz-ingestion-key': '<INGESTION_KEY>',

Reviewer 2: Also mention that ingestion key headers are optional for Self Hosted Version

    // Create a metric reader with OTLP exporter configured to send metrics to a local collector.
    const metricReader = new PeriodicExportingMetricReader({
      exporter: new OTLPMetricExporter({
        // For the self-hosted version, please use the collector url instead.
        url: `https://ingest.<INGESTION_REGION>.signoz.cloud:443/v1/metrics`,
        headers: {
          'signoz-ingestion-key': '<INGESTION_KEY>',
        },
      }),
      exportIntervalMillis: 10000, // Export metrics every 10 seconds.
    });

    // Initialize a MeterProvider with the above configurations.
    const myServiceMeterProvider = new MeterProvider({
      resource,
      readers: [metricReader],
    });

    // Set the initialized MeterProvider as global to enable metric collection across the app.
    metrics.setGlobalMeterProvider(myServiceMeterProvider);
    ```

    - Set the `<INGESTION_REGION>` to match your SigNoz Cloud [region](https://signoz.io/docs/ingestion/signoz-cloud/overview/#endpoint)
    - Replace `<INGESTION_KEY>` with your SigNoz [ingestion key](https://signoz.io/docs/ingestion/signoz-cloud/keys/)
    - `<SERVICE_NAME>` is the name of your service

  </TabItem>
</Tabs>

> <Admonition type="info">

Reviewer 2: Also add
```
update the endpoint and remove the ingestion key header as shown in
  <a href="/docs/ingestion/cloud-vs-self-hosted#cloud-to-self-hosted">Cloud → Self-Hosted</a>.
```

  For self-hosted deployments, add the following CORS configuration to your OpenTelemetry Collector
  config file to allow browser requests from your frontend domain:
</Admonition>

> ```yaml

Reviewer 2: Filename should be added as mentioned here - https://github.com/SigNoz/signoz.io/blob/78c082ce6475e6bf9253b0f4d9b4381377196383/CONTRIBUTING.md?plain=1#L117

Reviewer 2: Also add a link to otel config as mentioned here - https://github.com/SigNoz/signoz.io/blob/78c082ce6475e6bf9253b0f4d9b4381377196383/CONTRIBUTING.md?plain=1#L156

        }
        case 'FCP': {
          fcp.record(metric.value, customAttributes);
          break;
        }
        default: {
          console.log('unexpected metric name');
        }
      }
    }

    onCLS(sendToAnalytics);
    onINP(sendToAnalytics);
    onLCP(sendToAnalytics);
    onTTFB(sendToAnalytics);
    onFCP(sendToAnalytics);
    ```

  </TabItem>
</Tabs>

## Next Steps

### Setup Dashboard and alerts

<figure data-zoomable align="center">
  <img
    className="box-shadowed-image"
    src="/img/events/launch-week-1/web-vitals.webp"
    alt="Core Web Vitals Monitoring in SigNoz"
  />
  <figcaption>
    <i>Core Web Vitals Monitoring in SigNoz </i>
  </figcaption>
</figure>

SigNoz charts support thresholds with color coding. For example, the graph below shows three thresholds for `good`, `needs improvement`, and `critical` levels.

<figure data-zoomable align="center">
  <img
    className="box-shadowed-image"
    src="/img/events/launch-week-1/thresholds-chart.webp"
    alt="Thresholds in SigNoz"
  />
  <figcaption>
    <i>Thresholds provide a visual cue about your application's performance </i>
  </figcaption>
</figure>

### Out of Box Dashboard

> You can [import](https://signoz.io/docs/dashboards/import-dashboard/) our out-of-the-box [Dashboard for web vitals monitoring](https://raw.githubusercontent.com/SigNoz/dashboards/refs/heads/main/frontend-monitoring/web-vitals-monitoring-with-metrics.json)

Reviewer 2: Link this page instead of github link - https://signoz-web-git-sig-2668-signoz.vercel.app/docs/dashboards/dashboard-templates/frontend-monitoring/#web-vitals-with-metrics
